modules/host/storageobligations.go

```diff
@@ -235,35 +364,383 @@ func (h *Host) addStorageObligation(so *storageObligation) error {
+// multiple instances of the same virtual sector, the virtural sector will need
+// to appear in 'sectorsRemoved' multiple times. Same with 'sectorsGained'.
+func (h *Host) modifyStorageObligation(so *storageObligation, sectorsRemoved []crypto.Hash, sectorsGained []crypto.Hash, gainedSectorData [][]byte) error {
+	// Sanity check - obligation should be under lock while being modified.
+	soid := so.id()
+	_, exists := h.lockedStorageObligations[soid]
+	if !exists {
+		h.log.Critical("modifyStorageObligation called with an obligation that is not locked")
+	}
+	// Sanity check - there needs to be enough time to submit the file contract
+	// revision to the blockchain.
+	if so.expiration()-revisionSubmissionBuffer <= h.blockHeight {
+		h.log.Critical("revision submission window was not verified before trying to modify a storage obligation")
+		return errNoBuffer
+	}
+	// Sanity check - sectorsGained and gainedSectorData need to have the same length.
+	if len(sectorsGained) != len(gainedSectorData) {
+		h.log.Critical("modifying a revision with garbage sector data", len(sectorsGained), len(gainedSectorData))
+		return errInsaneStorageObligationRevision
+	}
+	// Sanity check - all of the sector data should be modules.SectorSize
+	for _, data := range gainedSectorData {
+		if uint64(len(data)) != modules.SectorSize {
+			h.log.Critical("modifying a revision with garbase sector sizes", len(data))
+			return errInsaneStorageObligationRevision
+		}
+	}
+
+	// Note, for safe error handling, the operation order should be: add
+	// sectors, update database, remove sectors. If the adding or update fails,
+	// the added sectors should be removed and the storage obligation shoud be
+	// considered invalid. If the removing fails, this is okay, it's ignored
+	// and left to consistency checks and user actions to fix (will reduce host
+	// capacity, but will not inhibit the host's ability to submit storage
+	// proofs)
+	var i int
+	var err error
+	for i = range sectorsGained {
+		err = h.addSector(sectorsGained[i], so.expiration(), gainedSectorData[i])
+		if err != nil {
+			break
+		}
+	}
+	if err != nil {
+		// Because there was an error, all of the sectors that got added need
+		// to be revered.
+		for j := 0; j < i; j++ {
+			// Error is not checkeed because there's nothing useful that can be
+			// done about an error.
+			_ = h.removeSector(sectorsGained[j], so.expiration())
+		}
+		return err
+	}
+	// Update the database to contain the new storage obligation.
+	err = h.db.Update(func(tx *bolt.Tx) error {
+		soBytes, err := json.Marshal(*so)
 		if err != nil {
 			return err
 		}
+		return tx.Bucket(bucketStorageObligations).Put(soid[:], soBytes)
+	})
+	if err != nil {
+		// Because there was an error, all of the sectors that got added need
+		// to be revered.
+		for i := range sectorsGained {
+			// Error is not checkeed because there's nothing useful that can be
+			// done about an error.
+			_ = h.removeSector(sectorsGained[i], so.expiration())
```

Comment:
actually, since you do this a lot, you could define a `revertSectors([]crypto.Hash)` helper function. That would mean fewer lines in the error branch above, which is always good.